lib/chef/resource/swap_file.rb

```diff
@@ -0,0 +1,217 @@
+#
+# Copyright 2012-2018, Seth Vargo
+# Copyright 2017-2018, Chef Software, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/resource"
+
+class Chef
+  class Resource
+    class SwapFile < Chef::Resource
+      resource_name :swap_file
+      provides :swap_file
+
+      description "Use the swap_file resource to create or delete swap files on Linux systems."
+      introduced "14.0"
+
+      property :path, String,
+               description: "The path to put the swap file on the system.",
+               name_property: true
+
+      property :size, Integer,
+               description: "The size (in MBs) for the swap file."
+
+      property :persist, [TrueClass, FalseClass],
+               description: "Persist the swapon.",
+               default: false
+
+      property :timeout, Integer,
+               description: "Timeout for dd/fallocate.",
+               default: 600
+
+      property :swappiness, Integer,
+               description: "The swappiness value to set on the system."
+
+      action :create do
+        description "Create a swapfile."
+
+        if swap_enabled?
+          Chef::Log.debug("#{new_resource} already created - nothing to do")
+        else
+          begin
+            Chef::Log.info "starting first create: #{node['virtualization']['system']}"
+            do_create(swap_creation_command)
+          rescue Mixlib::ShellOut::ShellCommandFailed => e
+            Chef::Log.warn("#{new_resource} Rescuing failed swapfile creation for #{new_resource.path}")
+            Chef::Log.debug("#{new_resource} Exception when creating swapfile #{new_resource.path}: #{e}")
+            do_create(dd_command)
+          end
+        end
+        if new_resource.swappiness
```

Comment:
I'm mixed on this feature. Notably if you have more than one swap, they would compete? But that's pretty rare.

Reply:
Separation of concerns suggests that this option shouldn't be here.  And it creates a really weird API where this is just a facade over calling the sysctl resource to set the swappiness.

Reply:
Two fold answer:

First part: It's an optional feature and it's off by default. It makes a lot of sense to manage your swapiness when you configure your swap file. For users that want to just get swap going they can set a value and it'll just work. If they don't turn it on, then they won't get it, which seems to follow "the chef way".

Second part: I think the scenario where we have multiple swap files and the users specify different swapiness values it super edge case. You could have multiple swap files and even set the swapiness on both without issue. You could specify it on one and it would work. You just can't have 2 different values and I think it *should* be pretty clear to anyone who would want swapiness set why that isn't possible.

Reply:
3rd part: Users of the cookbook are utilizing this property so removing it breaks that API in a sad way